File: packages/create-x402-service/templates/python-fastapi/app/x402/registry_client.py
import httpx
import logging
from typing import Optional, Dict, Any, List
from datetime import datetime

logger = logging.getLogger(__name__)


class RegistryClient:
    def __init__(self, registry_url: str):
        self.registry_url = registry_url
        self.service_id: Optional[str] = None
        self.client = httpx.AsyncClient(timeout=10.0)

    async def register_service(
        self,
        name: str,
        description: str,
        url: str,
        category: str,
        pricing: Dict[str, Any],
        wallet_address: str,
        network: str,
        accepted_tokens: List[str],
        capabilities: Optional[List[str]] = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        registration_data = {
            "name": name,
            "description": description,
            "url": url,
            "category": category,
            "pricing": pricing,
            "walletAddress": wallet_address,
            "network": network,
            "acceptedTokens": accepted_tokens,
            "capabilities": capabilities or [],
            "tags": tags or [],
            "metadata": metadata or {},
        }

        response = await self.client.post(f"{self.registry_url}/services/register", json=registration_data)
        response.raise_for_status()

        result = response.json()
        self.service_id = result.get("serviceId")
        logger.info(f"Service registered with x402 registry: {self.service_id}")

        return result.get("service", {})

    async def update_service(self, updates: Dict[str, Any]) -> Dict[str, Any]:
        if not self.service_id:
            raise ValueError("Service not registered")

        response = await self.client.put(f"{self.registry_url}/services/{self.service_id}", json=updates)
        response.raise_for_status()

        logger.info(f"Service updated in registry: {self.service_id}")
        return response.json().get("service", {})

    async def set_service_status(self, status: str) -> None:
        if not self.service_id:
            raise ValueError("Service not registered")

        await self.client.patch(f"{self.registry_url}/services/{self.service_id}/status", json={"status": status})
        logger.info(f"Service status updated: {self.service_id} -> {status}")

    async def report_metrics(self, metrics: Dict[str, Any]) -> None:
        if not self.service_id:
            raise ValueError("Service not registered")

        await self.client.post(f"{self.registry_url}/services/{self.service_id}/metrics", json=metrics)
        logger.debug(f"Metrics reported to registry: {self.service_id}")

    async def heartbeat(self) -> None:
        if not self.service_id:
            raise ValueError("Service not registered")

        await self.client.post(f"{self.registry_url}/services/{self.service_id}/heartbeat")
        logger.debug(f"Heartbeat sent to registry: {self.service_id}")

    async def get_service_info(self) -> Optional[Dict[str, Any]]:
        if not self.service_id:
            return None

        response = await self.client.get(f"{self.registry_url}/services/{self.service_id}")
        return response.json().get("service")

    async def close(self):
        await self.client.aclose()

File: packages/create-x402-service/templates/nestjs/src/x402/registry-client.ts
import axios from 'axios';
import pino from 'pino';

const logger = pino({ name: 'registry-client' });

export interface ServiceRegistration {
  url: string;
  name: string;
  description: string;
  category: string;
  ownerWalletAddress: string;
  pricePerCall: number;
  acceptedTokens: string[];
  capabilities?: string[];
  tags?: string[];
}

export interface Service {
  id: string;
  url: string;
  name: string;
  description: string;
  category: string;
  pricePerCall: number;
  acceptedTokens: string[];
  reputationScore: number;
  verified: boolean;
  uptime: number;
}

export class RegistryClient {
  private registryUrl: string;
  private serviceId?: string;

  constructor(registryUrl: string) {
    this.registryUrl = registryUrl;
  }

  async registerService(registration: ServiceRegistration): Promise<Service> {
    try {
      const response = await axios.post(
        `${this.registryUrl}/services/register`,
        registration
      );

      this.serviceId = response.data.serviceId;

      logger.info(
        { serviceId: this.serviceId, name: registration.name },
        'Service registered with x402 registry'
      );

      return response.data.service;
    } catch (error) {
      logger.error({ error }, 'Failed to register service');
      throw error;
    }
  }

  async updateServiceStatus(status: 'ACTIVE' | 'PAUSED' | 'DEPRECATED'): Promise<void> {
    if (!this.serviceId) {
      logger.warn('Service not registered, skipping status update');
      return;
    }

    try {
      await axios.patch(`${this.registryUrl}/services/${this.serviceId}`, {
        status,
      });

      logger.info({ serviceId: this.serviceId, status }, 'Service status updated');
    } catch (error) {
      logger.error({ error }, 'Failed to update service status');
    }
  }

  async reportMetrics(metrics: {
    responseTime: number;
    successRate: number;
  }): Promise<void> {
    if (!this.serviceId) {
      return;
    }

    try {
      await axios.post(`${this.registryUrl}/services/${this.serviceId}/metrics`, metrics);
    } catch (error) {
      logger.error({ error }, 'Failed to report metrics');
    }
  }

  async discover(filters?: {
    category?: string;
    maxPrice?: number;
    minReputation?: number;
  }): Promise<Service[]> {
    try {
      const response = await axios.get(`${this.registryUrl}/services/discover`, {
        params: filters,
      });

      return response.data.services;
    } catch (error) {
      logger.error({ error }, 'Failed to discover services');
      return [];
    }
  }

  getServiceId(): string | undefined {
    return this.serviceId;
  }
}
